src/network/netdev/macsec.c

```diff
@@ -0,0 +1,888 @@
+        if (r < 0)
+                return r;
+
+        dest = a ? &a->sa : &b->sa;
+
+        r = unhexmem(rvalue, strlen(rvalue), &p, &l);
+        if (r < 0) {
+                log_syntax(unit, LOG_ERR, filename, line, r,
+                           "Failed to parse key. Ignoring assignment: %m");
+                return 0;
+        }
+        if (l != 16) {
+                /* See DEFAULT_SAK_LEN in drivers/net/macsec.c */
+                log_syntax(unit, LOG_ERR, filename, line, r,
+                           "Invalid key length (%zu). Ignoring assignment", l);
+                return 0;
+        }
+
+        free_and_replace(dest->key, p);
+        dest->key_len = l;
+
+        TAKE_PTR(a);
+        TAKE_PTR(b);
+
+        return 0;
+}
+
+int config_parse_macsec_key_id(
+                const char *unit,
+                const char *filename,
+                unsigned line,
+                const char *section,
+                unsigned section_line,
+                const char *lvalue,
+                int ltype,
+                const char *rvalue,
+                void *data,
+                void *userdata) {
+
+        _cleanup_(macsec_transmit_association_free_or_set_invalidp) TransmitAssociation *a = NULL;
+        _cleanup_(macsec_receive_association_free_or_set_invalidp) ReceiveAssociation *b = NULL;
+        _cleanup_free_ void *p;
+        MACsec *s = userdata;
+        uint8_t *dest;
+        size_t l;
+        int r;
+
+        assert(filename);
+        assert(section);
+        assert(lvalue);
+        assert(rvalue);
+        assert(data);
+
+        if (streq(section, "MACsecTransmitAssociation"))
+                r = macsec_transmit_association_new_static(s, filename, section_line, &a);
+        else
+                r = macsec_receive_association_new_static(s, filename, section_line, &b);
+        if (r < 0)
+                return r;
+
+        r = unhexmem(rvalue, strlen(rvalue), &p, &l);
+        if (r < 0) {
+                log_syntax(unit, LOG_ERR, filename, line, r, "Failed to parse key id. Ignoring assignment: %s", rvalue);
+                return 0;
+        }
+        if (l > MACSEC_KEYID_LEN) {
+                log_syntax(unit, LOG_ERR, filename, line, 0,
+                           "The size of key id is too large (%zu), maximum of %zu permitted. "
+                           "Ignoring assignment: %s", l, (size_t) MACSEC_KEYID_LEN, rvalue);
+                return 0;
+        }
+
+        dest = a ? a->sa.key_id : b->sa.key_id;
+        memcpy_safe(dest, p, l);
+        memzero(dest + l, MACSEC_KEYID_LEN - l);
+
+        TAKE_PTR(a);
+        TAKE_PTR(b);
+
+        return 0;
+}
+
+static int macsec_receive_channel_verify(ReceiveChannel *c) {
+        NetDev *netdev;
+
+        assert(c);
+        assert(c->mac_sec);
+
+        netdev = NETDEV(c->mac_sec);
+
+        if (section_is_invalid(c->section))
+                return -EINVAL;
+
+        if (eqzero((const void *) &c->sci.mac))
```

Comment:
this check is not right. you are comparing the ptr by value against zero, not the contens

Reply:
fixed.